Open pull request to discuss and review the potential changes

cklanac opened the pull request.

tomatau commented:
Oh we're not giving this project to students, they'll have gone through the Node content first and have built their own version of this project: https://github.com/tomatau/express-boilerplate

There are a few things missing from this project too such as:
- API_KEY checks
- XSS sanitising responses
- specifying the fields in responses

cklanac commented:
I like the services approach. I have something similar using `queries.js` files in my project.

- Recommend using integers for primary keys instead of UUIDs. The UUID are not strictly necessary and will likely cause some issues when keying in ID for Postman.
- Recommend removing the 'blogful_' prefix for the table names. 
- Prefer using the plural for table name (articles, comments, users, tags)
- Removed Procfile. Heroku calls `npm start` so the procfile is not necessary
- Move database reference from `req.db` to a module. 
    - Modules are cached so subsequent requires will not create a new connection
    - Yes, the queries are no longer pure functions but some of them reference `this.`. Plus express endpoints don't lend themselves to pure functions so our software design is mixed.
    - If dependency injection is the goal then I suggest we inject the DB reference into the module.
- Move `cors` and `logger` to middleware. It helps re-affirm the middleware pattern and provides a nice transition for simple middleware earlier in the course.
- Changed the endpoints to plural `/articles`, `/tags` etc
- Recommend prefixing the endpoints with `/api/` and possibly version (e.g `/api/v1/`)
- Changed the name of the Service files to lowercase and just `service.js` to match "organize-by-feature" pattern. If you prefer using verbose filename like `passport-jwt.controller.js` then
    - Prefer file names to be lowercase. 
    - The name of the file should be lowercase and skewered.
    - The file type (service, controller, router) should be a dot extension e.g. `/routers/

tomatau commented:
> Recommend using integers for primary keys instead of UUIDs. The UUID are not strictly necessary and will likely cause some issues when keying in ID for Postman.

Keying issue? Integers are very insecure and easily guessable, we're trying to focus on security. We've also used UUIDs a lot on previous content.

UUIDs also make it easier to test as there are deterministic.

> Recommend removing the 'blogful_' prefix for the table names.

Why? 
The intention is an alternative to a schema and an extra security measure as less guessable. Also explicit with namespacing

> Prefer using the plural for table name (articles, comments, users, tags)

Why? [Check out this SO for reasoning on singular](https://stackoverflow.com/questions/338156/table-naming-dilemma-singular-vs-plural-names), it also discusses prefixing

> Removed Procfile. Heroku calls npm start so the procfile is not necessary

This is to keep the command specific for Heroku. Avoids risk of people updating their `npm start` to `nodemon` and not realising it hurts production app. Beside that main point, 2 extra things: using `npm` has a little overhead ;) and you can add extra conditions specific to production if necessary.

> Move database reference from req.db to a module.

The goal is for DI and easier testing when we introduce these service files

> If dependency injection is the goal then I suggest we inject the DB reference into the module.

This would require creating "instances" of the service and make them more complicated. Or doing some `.setDB` method which is a very stateful paradigm that makes me shudder a bit.

> Move cors and logger to middleware. It helps re-affirm the middleware pattern and provides a nice transition for simple middleware earlier in the course.

I find it a bit of a layer of obscurity and indirection. Not sure how it re-affirms other than using the word middleware?

> Changed the endpoints to plural /articles, /tags etc

Using a plural was for the same reason in the table name, also `/tags/:tag_id` always feels weird to me as it's returning an individual entity. I'm happy to change this one though, feels quite subjective.

> Recommend prefixing the endpoints with /api/ and possibly version (e.g /api/v1/)

I'd be happy to do this, it would require more content to explain the versioning and also the whole app is an API so not sure what the benefit is in adding the extra verbosity?

cklanac commented:
My intent with most of these suggestions is to simply the sample app. Nothing that isn't strictly necessary. I know that's subjective but at least you know my motivation.

- UUIDs, Yes, a keying issue. We see it all the time in mongo? And I thought postgres UUIDs were sequential, but it looks like you're using `uuid_generate_v4()` which generates random ids

- Removing 'blogful_' prefix for the table names. This is security through obscurity and just makes the code unnecessarily verbose. See this ref:
https://security.stackexchange.com/questions/119510/is-using-a-db-prefix-for-tables-more-secure

- Prefer using the plural for table name (articles, comments, users, tags). Why? Probably just because that's the way I learned it. I think common practice has changed (mutters to self... "I'm old, damn kids!")

- Remove Procfile. I think adding the Procfile can also cause potential confusion and mistakes. So at best it's a draw. I still prefer to remove it. And only add it in if there is a pressing need.

- The DI that I'm referring to is just passing in a reference to the DB in the require.
```
// in a test file
const knex = knex({
  client: 'pg',
  connection: TEST_DB_URL,
})
const ArticleService = require(`../article/service`)(knex)
...
```

Middleware - the comment assumes students created a custom logger like 
```
const simpleLogger = function (req, res, next) {
  const now = new Date();
  console.log(`${now.toLocaleString()} ${req.method} ${req.url}`);
  next();
};

module.exports = simpleLogger;
```

Then later, they swap out the custom logger for morgan. The module also has the advantage of keeping the `app.js` a bit cleaner as more configurations are added to cors and logger. For instance, if security is your concerns then you'll want to add a CLIENT_ORIGIN to cors 

```
const cors = require('cors');

const { CLIENT_ORIGIN } = require('../config');

module.exports = cors({ origin: CLIENT_ORIGIN });
```

Plural resource endpoints: It is the standard, not an opinion.
https://github.com/byrondover/api-guidelines/blob/master/Guidelines.md#url-structure
https://apigee.com/about/blog/technology/restful-api-design-plural-nouns-and-concrete-names
https://blog.mwaysolutions.com/2014/06/05/10-best-practices-for-better-restful-api/
etc...

Prefix with `/api/v1` - Typically considered best practice (see REST links above) but I'm not strongly opinionated about it.

I also still have my concerns about using `COALESCE, JSON_AGG , TO_JSON...`. We used to use Treeize to convert tabular data to an object tree.
https://www.npmjs.com/package/treeize

tomatau commented:
Finding the balance between simple content, simple code, giving guidance that's secure and scalable is the tricky thing. I think everyone prefers more simple and often that's the difficult skill people acquire with more coding experience; and then yes, simple for one person is the opposite of simple for another depending on how each learned it the first time.

- AFAIK postgres doesn't support UUIDs without an extension.
- Yes exactly security through obscurity! I'm OK with removing it but security through obscurity is still valuable.
- The Procfile is already in the boilerplate they clone for new projects so it's not really extra work for students and opens up the discussed potential.
- Yeah that DI is the DI I was talking about as well, you need to manage an instance of the service that's created at each `require()` when you invoke and you also need to pass the db instance around so you can inject it as well, it gets messy! Especially messy if you use the service in multiple files, e.g. have services using services. 

Method level DI is easier as the Service is already required and available for the middleware. The db instance can be attached to the request or app itself as you can defer injection without creating instances. 

We can completely remove the DI stuff and just have the db instance required into services like your PR has it. It's not my flavour for the reasons I mentioned in above comment ^ but if it's the consensus then I'll go with it!

(gonna remove the references to `this` inside services even though they aren't harmful here, just on principle)
- I can't remember anywhere we make a custom logger in the new content. I'm all for making the `app.js` file cleaner and the client_origin is definitely something we'll introduce when it becomes relevant at the full stack point I'm sure! My preference when building these apps is to keep the app's requirements clear in the `app.js` file and to modularize and layer custom code but yeh it's subjective stuff.
- The plural endpoints still fall victim to the language, consistency and convenience arguments I linked to for plural table names.  I think the popular option is still "[go with what you like](https://stackoverflow.com/questions/6845772/rest-uri-convention-singular-or-plural-name-of-resource-while-creating-it)" on this topic. (I don't see any official standards here?) 

Let's do both! 

```js
app
  .use(/\/cherry$|\/cherries$/, cherryRouter)
  .use(/\/goose$|\/geese$/, gooseRouter)
```
(joking)
- The coalesce is only there to highlight alternative solutions and capabilities of Postgres! Not going to dive into it other than label it as "advanced". Working with these functions grows super complicated with many to many relationships and would be waaaaay too complicated for students!

tomatau commented:
How do you feel about what's returned from the endpoints?

e.g. `POST /article/:article_id/tag` 
- returns the array of tags for the article.
- instead of the tag, as the tag wasn't created here, only the relationship.
- could return the full modified article with new tag, feels a bit strange to do this though?

PATCH is returning "no content"s

How do you feel about the knex queries we're using? E.g. 4 queries to get all the articles and then build the collections with nested relationships. This has scaling issues of course! Might be better options? Could say we definitely go with bookshelfJS at this point or sequelize.

cklanac commented:
- UUIDs - Still prefer integers for students, but willing to live w/ `uuid_generate_v4()`
- Table prefix - Please remove the table prefixes
- Plural table names - Prefer plural but willing to go with singular names
- Procfile - It still seems unnecessary. Please remove it.
- DB reference - Modules are cached, so if you load the connection from a module, then you don't need to worry about passing the reference around.
```
const knex = require('../knex); //
const ArticleService = require(`../article/service`)(knex)
```
See: https://github.com/tomatau/blogful-api/pull/1/commits/2390b1670579543dba9e5b5e7ef5e8298bde0564#diff-998928d3dfd66a561e9605c44a436ce4

- Middleware as modules - Strongly prefer the modularized middleware. It ties into custom middleware for authentication. And it's not that obscure.
- Plural endpoints - REST is not a spec, it's an architectural style. That said, there are conventions, please pluralize the endpoints
- Endpoint prefix - It is considered a best practice but I'm OK with keeping simplifying
- Organize by feature and file names - file names lowercase and skewered. Use `.type.js` to denote special files like `file-name.model.js` and `file-name.service.js`
- Nested resources - Nested resources are fine. POST doesn't need to return the new item. Just the location with a newId e.g. `.location('/path/to/resource/newID)`
- PATCH return 204 no content. Then we should be consistent. POST, PATCH (PUT) and DELETE should all return no content.

alfaraday commented:
Changes to make:

- [x] THT: Pluralize endpoints and table names
- [x] THT: Add API endpoint prefixes
- [ ] AF: Decide on procfile in review



AF & CK to discuss the rest

alfaraday commented:
Will do customized CORS in the full stack deployment checkpoint with Noteful because the blog app doesn't have a client at the moment.

alfaraday commented:
* **Middleware as modules:** The actual concern here is finding a gentle introduction to the services pattern. We could use Morgan or error-handling as an early example. This is something we should address in the earlier checkpoints. We don't really need to move Morgan or CORS in the final version of the blog app.
* **Table prefixes:** We're keeping them. More verbose, but can prevent greater pain down the road.
* **PATCH/POST:** PATCH and POST responses should have the same status. If that's 204, send the id for POST in a location header. But it doesn't need to be 204.

cklanac commented:
> PATCH/POST: PATCH and POST responses should have the same status. If that's 204, send the id for POST in a location header. But it doesn't need to be 204.

When creating the previous version of Noteful I used GitHub API V3 as a guide. When in doubt I followed their lead. For example: https://developer.github.com/v3/issues

POST return 200 with location and body with results
PATCH return 200 and body with results
DELETE return 204 No Content

tomatau commented:
I'm fine with using 200s but there's value in using 201. We've also used 201 in other points in the node curriculum.

The [specification for POST](https://tools.ietf.org/html/rfc7231#section-4.3.3) states you should use 201 as well. I'm sure we can find lots of examples of existing APIs that do all kinds of things.

(Also according to spec: PATCH and PUT are able to supply 201 if they decide to create the resource that didn't exist rather than responding a 404. It should be clear that endpoints don't need to all send the same status, the endpoints do different things and one endpoint can deal with multiple statuses)
